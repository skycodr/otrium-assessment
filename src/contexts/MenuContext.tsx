import { createContext } from "react";
import { useFacets } from "../hooks";

import json from '../__data__/categories.json';

export const MenuContext = createContext<ITreeContext<ICategory>>({
  dataSource: [],
  getNodes: (_facet) => [],
  setNodes: (_facet, _value) => [],
  commit: () => { }
});

const MenuContextProvider = ({ children }: IChildren) => {
  const [
    dataSource,
    getNodes,
    setNodes,
    commit
  ] = useFacets<ICategory>(json.data.categories);

  return (
    <MenuContext.Provider value={{ dataSource, getNodes, setNodes, commit }}>
      {children}
    </MenuContext.Provider>
  );
};

export default MenuContextProvider;