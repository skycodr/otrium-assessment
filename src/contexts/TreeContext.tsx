import { createContext } from "react";
import { useFacets } from "../hooks";

import json from '../__data__/categories.json';

// Context Object
export const TreeContext = createContext<ITreeFacets<ICategory>>({
  dataSource: [],
  getFacets: (_facet) => [],
  setFacets: (_facet, _value) => [],
  removeFacets: (_facet) => { },
  commit: () => { }
});

// Context Provider
const TreeContextProvider = ({ children }: IChildren) => {
  const {
    dataSource,
    getFacets,
    setFacets,
    removeFacets,
    commit
  } = useFacets<ICategory>(json.data.categories);

  return (
    <TreeContext.Provider value={{ dataSource, getFacets, setFacets, removeFacets, commit }}>
      {children}
    </TreeContext.Provider>
  );
};

export default TreeContextProvider;