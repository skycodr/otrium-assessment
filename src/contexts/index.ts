export {default as MenuContextProvider, MenuContext } from './MenuContext';