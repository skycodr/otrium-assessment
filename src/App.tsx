import { Checkbox, FormControlLabel, Grid, makeStyles } from '@material-ui/core';
import { useContext } from 'react';
import { Leaf, Tree } from './components';
import { TreeMenuContext } from './contexts';
import { useTree } from './hooks';
import { getCategoryKey, getCategoryLabel, getCategoryChildFacet } from './utils';

const useTreeStyles = makeStyles({
  treeStyle: {
    flexGrow: 1,
    maxWidth: 400,
  },
});


const App = () => {

  const styles = useTreeStyles();
  const { getNodes, handleSelect, toggleSelectAll } = useTree<ICategory>(
    useContext(TreeMenuContext), 
    getCategoryChildFacet);

  /**
   * Special Note:
   * 
   * If you want to get all selected categories. Simply write,
   *  getNodes({checked: true});
   */

  /*
   * Dev Note:
   *
   *  I have written some confusing logic on purpose. :D. Normally wouldn't do this
   *  on a project. Such as, bunching up things together without first using variables etc.
   */

  return (
    <Grid direction='column' container>
      <Grid>
        <FormControlLabel 
          label='Toggle select all' 
          control={<Checkbox onChange={(e) => toggleSelectAll(e.target.checked)} />} />
      </Grid>
      <Grid>
        <Tree<ICategory>
          className={styles.treeStyle}
          root={{ parent: '0' }}
          sourceExtractor={getNodes}
          keyExtractor={getCategoryKey}
          rootExtractor={getCategoryChildFacet}
          // You may find branches are on the tree ;)
          render={(value: ICategory) => (
            <Leaf label={getCategoryLabel(value)} onChange={(e) => {
              e.stopPropagation();
              handleSelect(value, e.target.checked);
            }} checked={!!value.checked} />
          )}
        />
      </Grid>
    </Grid>
  );
};

export default App;
