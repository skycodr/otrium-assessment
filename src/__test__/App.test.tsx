import { render, screen } from '@testing-library/react';
import App from '../App';
import { MenuContextProvider } from '../contexts';

describe('App initial state tests', () => {
  test('App renders renders properly', () => {
    const { asFragment } = render(
      <MenuContextProvider>
        <App />
      </MenuContextProvider>
    );

    const selectAllElement = screen.getByText(/Toggle select all/);

    expect(asFragment()).toMatchSnapshot();
    expect(selectAllElement).toBeInTheDocument();

  });
});
