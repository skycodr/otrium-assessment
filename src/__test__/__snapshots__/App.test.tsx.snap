// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App initial state tests App renders renders properly 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column"
  >
    <div
      class="MuiGrid-root"
    >
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-2 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
          data-testid="selectAll"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              class="PrivateSwitchBase-input-5"
              data-indeterminate="false"
              type="checkbox"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Toggle select all
        </span>
      </label>
    </div>
    <div
      class="MuiGrid-root"
    >
      <ul
        aria-multiselectable="false"
        class="MuiTreeView-root makeStyles-treeStyle-1"
        role="tree"
      >
        <li
          aria-expanded="false"
          class="MuiTreeItem-root"
          role="treeitem"
          tabindex="-1"
        >
          <div
            class="MuiTreeItem-content"
          >
            <div
              class="MuiTreeItem-iconContainer"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                />
              </svg>
            </div>
            <div
              class="MuiTypography-root MuiTreeItem-label MuiTypography-body1"
            >
              <div
                class="MuiBox-root MuiBox-root-6"
              >
                <span
                  aria-disabled="false"
                  class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-2 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                  data-testid="treeItem_14100"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <input
                      class="PrivateSwitchBase-input-5"
                      data-indeterminate="false"
                      type="checkbox"
                      value=""
                    />
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </span>
                Dames (136)
              </div>
            </div>
          </div>
        </li>
        <li
          aria-expanded="false"
          class="MuiTreeItem-root"
          role="treeitem"
          tabindex="-1"
        >
          <div
            class="MuiTreeItem-content"
          >
            <div
              class="MuiTreeItem-iconContainer"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                />
              </svg>
            </div>
            <div
              class="MuiTypography-root MuiTreeItem-label MuiTypography-body1"
            >
              <div
                class="MuiBox-root MuiBox-root-7"
              >
                <span
                  aria-disabled="false"
                  class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-2 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                  data-testid="treeItem_21251"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <input
                      class="PrivateSwitchBase-input-5"
                      data-indeterminate="false"
                      type="checkbox"
                      value=""
                    />
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </span>
                Heren (14)
              </div>
            </div>
          </div>
        </li>
        <li
          aria-expanded="false"
          class="MuiTreeItem-root"
          role="treeitem"
          tabindex="-1"
        >
          <div
            class="MuiTreeItem-content"
          >
            <div
              class="MuiTreeItem-iconContainer"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                />
              </svg>
            </div>
            <div
              class="MuiTypography-root MuiTreeItem-label MuiTypography-body1"
            >
              <div
                class="MuiBox-root MuiBox-root-8"
              >
                <span
                  aria-disabled="false"
                  class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-2 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
                  data-testid="treeItem_14126"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <input
                      class="PrivateSwitchBase-input-5"
                      data-indeterminate="false"
                      type="checkbox"
                      value=""
                    />
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </span>
                Kids (154)
              </div>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
